contracts/transfer/src/state.rs

```diff
@@ -0,0 +1,651 @@
+        // 16. N_p^*←encode(N_p^t)
+        // 17. N↦.append((N_p^t.R, N_p^t.pk))
+        // 18. Notes.append(N_p^*)
+        self.push_fee_crossover(tx.fee)
+            .expect("Failed to append the fee and the crossover to the state!");
+
+        self.update_root();
+
+        res
+    }
+
+    /// Push a note to the contract's state with the given block height
+    ///
+    /// Note: the method `update_root` needs to be called after the last note is
+    /// pushed.
+    pub fn push_note(&mut self, block_height: u64, note: Note) -> Note {
+        let pos = self.tree.push(TreeLeaf { block_height, note });
+        self.get_note(pos)
+            .expect("There should be a note that was just inserted")
+    }
+
+    /// Return the leaves in a given block height range.
+    pub fn leaves_in_range(&self, range: Range<u64>) -> Vec<TreeLeaf> {
+        match self.tree.leaves(range) {
+            Some(leaves) => leaves.cloned().collect(),
+            None => vec![],
+        }
+    }
+
+    /// Update the root for of the tree.
+    pub fn update_root(&mut self) -> BlsScalar {
+        let root = self.tree.root();
+        self.roots.insert(root);
+        root
+    }
+
+    /// Get the root of the tree.
+    pub fn root(&self) -> BlsScalar {
+        self.tree.root()
+    }
+
+    /// Get the opening
+    pub fn opening(
+        &self,
+        pos: u64,
+    ) -> Option<PoseidonBranch<TRANSFER_TREE_DEPTH>> {
+        self.tree.opening(pos)
+    }
+
+    /// Takes some nullifiers and returns a vector containing the ones that
+    /// already exists in the contract
+    pub fn existing_nullifiers(
+        &self,
+        nullifiers: Vec<BlsScalar>,
+    ) -> Vec<BlsScalar> {
+        nullifiers
+            .into_iter()
+            .filter_map(|n| self.nullifiers.get(&n).map(|_| n))
+            .collect()
+    }
+
+    /// Return the balance of a given contract.
+    pub fn balance(&self, module_id: &ModuleId) -> u64 {
+        self.balances.get(module_id).copied().unwrap_or_default()
+    }
+
+    /// Add balance to the given contract
+    pub fn add_balance(&mut self, module: ModuleId, value: u64) {
+        match self.balances.entry(module) {
+            Entry::Vacant(ve) => {
+                ve.insert(value);
+            }
+            Entry::Occupied(mut oe) => {
+                let v = oe.get_mut();
+                *v += value
+            }
+        }
+    }
+
+    pub fn message(
+        &self,
+        module: &ModuleId,
+        pk: &PublicKey,
+    ) -> Option<Message> {
+        let map = self.message_mapping.get(module)?;
+        let message = map.get(&pk.to_bytes())?;
+
+        Some(*message)
+    }
+
+    fn get_note(&self, pos: u64) -> Option<Note> {
+        self.tree.get(pos).map(|l| l.note)
+    }
+
+    fn any_nullifier_exists(&self, nullifiers: &[BlsScalar]) -> bool {
+        nullifiers
+            .iter()
+            .fold(false, |t, n| t || self.nullifiers.get(n).is_some())
```

Comment:
Is this short-circuiting in case the first nullifier exists?

Reply:
The compiler should take care of it, but to ensure it does - and to make it clearer - I changed it.